crates/storage-pg/src/upstream_oauth2/link.rs

```diff
@@ -355,4 +355,50 @@ impl<'c> UpstreamOAuthLinkRepository for PgUpstreamOAuthLinkRepository<'c> {
             .try_into()
             .map_err(DatabaseError::to_invalid_operation)
     }
+
+    #[tracing::instrument(
+        name = "db.upstream_oauth_link.remove",
+        skip_all,
+        fields(
+            db.query.text,
+            upstream_oauth_link.id,
+            upstream_oauth_link.provider_id,
+            %upstream_oauth_link.subject,
+        ),
+        err,
+    )]
+    async fn remove(&mut self, upstream_oauth_link: UpstreamOAuthLink) -> Result<(), Self::Error> {
+        let mut tx = self.conn.begin().await?;
+
+        // Unset the authorization sessions first, as they have a foreign key
+        // constraint on the links.
+        sqlx::query!(
+            r#"
+                    UPDATE upstream_oauth_authorization_sessions SET upstream_oauth_link_id = NULL
```

Comment:
It looks like I modelled the link on the rust side to not be possible to be in this state:

https://github.com/element-hq/matrix-authentication-service/blob/44c31ed142cbbc82fba55bfda74357b9baa17c80/crates/data-model/src/upstream_oauth2/session.rs#L23-L28

I think we should:

 - add an 'unlinked_at' timestamp
 - add a variant to this enum like
   ```rust
   Unlinked {
     competed_at: DateTime<Utc>,
     consumed_at: Option<DateTime<Utc>>,
     unlinked_at: DateTime<Utc>,
     id_token: Option<String>
   }
   ```
 - set this timestamp when we null this. This also means passing the clock to this `remove` function

Could you split out the 'link removal' API in a separate PR? That would make it easier to iterate on it, add tests, etc.